syntax = "proto2";
package sgfc;

message FlightMessage {
  optional string sender = 1;
  optional string auth_key = 2;
  optional string sequence_id = 3;

  repeated Payload payload = 1000;
}

message Payload {
  required PayloadType type = 1;

  oneof data {
    GpsPosition gps_position = 100;
    FlightControlCommand flight_control_command = 101;
    CameraData camera_data = 102;
  }
}

enum PayloadType {
  GPS_POSITION = 1;
  FLIGHT_CONTROL_COMMAND = 2;
  CAMERA_IMAGE = 3;
}

enum CameraDataType {
  RAW = 1;
  JPEG = 2;
}

message GpsPosition {
  optional bool has_fix = 1;
  optional double latitude = 2;
  optional double longitude = 3;
  optional double altitude = 4;
  optional double speed = 5;
}

message CameraData {
  optional string source_id = 1;
  optional CameraDataType type = 2;
  optional bytes data = 3;
}

message FlightControlCommand {
  optional double longitudinal = 1;
  optional double lateral = 2;
  optional double pitch = 3;
  optional double roll = 4;
  optional double yaw = 5;
  optional double throttle = 6;
}
